### ISSUE TITLE
[QUESTION] Why are hooks not rendered?

### ISSUE THREAD

**Issue by james-callahan** [mgoltzsche/khelm#16]
When running under khelm, why are hooks not rendered?
```
$ helm template kube-prometheus-stack --repo=https://prometheus-community.github.io/helm-charts | grep helm.sh/hook
    "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": test-success
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
```
vs
```
$ khelm template kube-prometheus-stack --repo=https://prometheus-community.github.io/helm-charts | grep helm.sh/hook
Running khelm 2.0.0-beta.4-3bad288 (helm 3.5.2)
Using repository "https://prometheus-community.github.io/helm-charts" (unauthenticated)
Downloading repository index of https://prometheus-community.github.io/helm-charts
Using chart kube-prometheus-stack from cache at /home/james/.cache/helm/repository/khelm/github.com/prometheus-community/helm-charts/releases/download/kube-prometheus-stack-14.6.2/kube-prometheus-stack-14.6.2-877bca9571978a86/kube-prometheus-stack-14.6.2.tgz
Rendering chart kube-prometheus-stack 14.6.2 with name "release-name" and namespace ""
```

**Comment by mgoltzsche** [comment]
@james-callahan Thanks for reporting the issue!

I am afraid currently khelm doesn't support Helm hooks.
khelm is a pure rendering tool. In opposite to a deployer like `helm install` it doesn't have control over the lifecycle of the deployment which is managed by another tool like `kpt live apply` (and this project's primary goal is to have a tool that integrates helm into generic, modular, declarative GitOps-enabled workflows where helm templates are rendered independent of a particular k8s cluster). Therefore translating Helm hooks into a static manifest is problematic since they are designed to be deployed during specific phases of the deployment before/after certain other resources are deployed and ready and their resources eventually removed afterwards as I understand.

Though the prometheus chart really seems to rely on those hooks because they contain Jobs that e.g. generate a TLS secret for a webhook.

As I understand hooks are used for the following use cases:
1) explicitly specifying an order in which resources are applied to the cluster.
2) waiting for each group of resources (e.g. initialization Jobs) to become available before applying the next group (when `--wait` is provided to `helm install`).
3) Running test Jobs after the installation.
4) Running clean up Jobs on deinstallation.

Eventually khelm could also output these hooks in order to support 1). For some charts this maybe even enough to support 2) but certainly not for any random chart. Also a k8s Job's spec is immutable so that updates wouldn't work this way. To support 3) and 4) I figure that resources for each hook could be wrapped in a special deployment resource that emulates the helm behaviour within the cluster which however would come with additional complexity and would make it hard to kustomize such a helm chart's output.

If you have ideas how to make khelm support Helm hooks or more use cases that should be considered please let me know.

**Comment by james-callahan** [comment]
> er). Correspondingly translating Helm hooks into a declarative static manifest is problematic since hooks are designed to be deployed during specific phases of the deployment before/after certain other resources are deployed and ready and their resources eventually removed afterwards as I understand while, when deploying a static manifest using e.g. kubectl, all resources are deployed at once independent of their initialization order.

I use ArgoCD which interprets the helm hooks (as emitted as annotations by e.g. `helm template`) into things that need to be ran at the relevant times; see https://argoproj.github.io/argo-cd/user-guide/helm/#helm-hooks

**Comment by mgoltzsche** [comment]
Interesting! Consequently the `khelm template` command should behave like `helm template` by default indeed.
At least for the kustomize exec and container plugin the hook behaviour should be configurable and we could introduce more hook handling strategies later.

**Comment by mgoltzsche** [comment]
Apparently making `khelm template` also print install hooks is as simple as enabling `IsInstall` in Helm's `ReleaseOptions` struct [here](https://github.com/mgoltzsche/khelm/blob/0f6ef1822901a715e7111f8223c10791f0f07902/pkg/helm/render.go#L65).
Of course a corresponding e2e test would need to be added as well - ideally using a minimal local chart that specifies a resource for each supported Helm hook. The PR should target the master branch to add the feature to khelm's Helm 2 support. I'd merge it to `v2` afterwards.
Anybody up for a PR?

**Comment by mgoltzsche** [comment]
ah, while Helm 2 renders all hooks (but there was no option to omit them) in Helm 3 there's additional code in the `template` command to explicitly render hooks as well if it isn't disabled - this code is missing in khelm 2.

**Comment by mgoltzsche** [comment]
@james-callahan should be fixed with khelm 2.0.0-beta.6

**Comment by james-callahan** [comment]
awesome, thank you!
